Operator + Forwarder docs (#1101)
smartcontractkit/documentation: src/pages/chainlink-nodes/v1/using-forwarder.md

- Allow the two Chainlink node EOAs to call the Forwarder's [forward](/chainlink-nodes/contracts/forwarder#forward) function. Because the Operator contract owns the Forwarder contract, you will have to call [acceptAuthorizedReceivers](/chainlink-nodes/contracts/operator#acceptauthorizedreceivers) on the Operator contract with the Forwarder contract address and the two Chainlink node EOAs as parameters. This call will make the Operator contract accept ownership of the Forwarder contract and authorize the Chainlink node EOAs to call the Forwarder contract by calling [setauthorizedsenders](/chainlink-nodes/contracts/receiver#setauthorizedsenders).

### Whitelist the Forwarder

In the blockchain explorer (Operator contract page), call `setAuthorizedSenders` with the address of your Forwarder contract. The parameter is an array. E.g., `["0xa3a10AB6A41eA19637Bb21b7decDA2d77138DfBa"]`. Metamask prompts you to confirm the transaction.

<ClickToZoom src='/images/chainlink-nodes/node-operators/forwarder/operatorsetauthorizedsenders.jpg' />

### Whitelist the Chainlink node EOAs

In the blockchain explorer (Operator contract page), call `acceptAuthorizedReceivers` with:

> - The Forwarder address in `targets`. E.g., `["0xa3a10AB6A41eA19637Bb21b7decDA2d77138DfBa"]`
> - The two Chainlink node EOAs in `sender`. E.g., `["0x0B87B9Ab848C729F152D87bBEbD4B9708757a59F","0x3218AB3Bf9B2AA2d440a9Ee2c773d1D4FC703Ef1"]`.

Reviewer: ```suggestion
- `targets`: Specify an array of forwarder addresses. For example, `["0xa3a10AB6A41eA19637Bb21b7decDA2d77138DfBa"]`
- `sender`: Specify an array with the two Chainlink node EOAs. For example, `["0x0B87B9Ab848C729F152D87bBEbD4B9708757a59F","0x3218AB3Bf9B2AA2d440a9Ee2c773d1D4FC703Ef1"]`.
```

Metamask prompts you to confirm the transaction.

<ClickToZoom src='/images/chainlink-nodes/node-operators/forwarder/acceptAuthorizedReceivers.jpg' />

## Create directRequest Jobs

This section is similar to the [Fulfilling Requests](/chainlink-nodes/v1/fulfilling-requests) guide.

1. In the Chainlink Operator UI on the **Jobs** tab, click **New Job**.
1. Create the `uint256` job. Replace:

>    - `YOUR_OPERATOR_CONTRACT_ADDRESS` with the address of your deployed Operator contract address.

Reviewer: ```suggestion
   - `YOUR_OPERATOR_CONTRACT_ADDRESS` with the address of your deployed operator contract address.
```

   - `EOA_ADDRESS` by the **first** Chainlink node EOA.

     <CodeSample src="samples/ChainlinkNodes/forwarder/get-uint256.toml"/>

1. Create the `string` job. Replace:

>    - `YOUR_OPERATOR_CONTRACT_ADDRESS` with the address of your deployed Operator contract address.

Reviewer: ```suggestion
   - `YOUR_OPERATOR_CONTRACT_ADDRESS` with the address of your deployed operator contract address.
```

   - `EOA_ADDRESS` with the **second** Chainlink node EOA.

     <CodeSample src="samples/ChainlinkNodes/forwarder/get-string.toml"/>

> 1. After creation, note the two Job Ids.

Reviewer: ```suggestion
1. After you create the jobs, record the two job IDs.
```

:::note

> Note both jobs have the attribute `forwardingAllowed = true`. This attribute will make the jobs attempt to run through the forwarder before falling back to default mode (e.g., if the Chainlink node tracks no forwarder).

Reviewer: ```suggestion
Note that both jobs have the attribute `forwardingAllowed = true`. This attribute will make the jobs attempt to run through the forwarder before falling back to default mode. For example, if the Chainlink node tracks no forwarder, the job falls back to the default mode.
```

:::

## Test the transaction-sending strategy

### Create API Requests

> 1. Open [APIConsumerForwarder.sol](https://remix.ethereum.org/#url=https://docs.chain.link/samples/APIRequests/APIConsumerForwarder.sol) in RemixIDE.

Reviewer: ```suggestion
1. Open [APIConsumerForwarder.sol](https://remix.ethereum.org/#url=https://docs.chain.link/samples/APIRequests/APIConsumerForwarder.sol) in the Remix IDE.
```

1. Note that `setChainlinkToken(0x326C977E6efc84E512bB9C30f76E30c160eD06FB)` is configured for _Goerli_.

1. On the **Compiler** tab, click the **Compile** button for `APIConsumerForwarder.sol`.

1. On the **Deploy and Run** tab, configure the following settings:

   - Select _Injected Provider_ as your environment. Make sure your metamask is connected to Goerli.
   - Select _APIConsumerForwarder_ from the **Contract** menu.

1. Click **Deploy**. MetaMask prompts you to confirm the transaction.

> 1. Fund the contract by sending LINK to the contract's address. See the [Fund your contract](/resources/fund-your-contract/) page for instructions. The address for the `ATestnetConsumer` contract is on the list of your deployed contracts in Remix. You can fund your contract with 1 LINK.

Reviewer: ```suggestion
1. Fund the contract by sending LINK to the contract's address. See the [Fund Your Contracts](/resources/fund-your-contract/) page for instructions. The address for the `ATestnetConsumer` contract is on the list of your deployed contracts in Remix. You can fund your contract with 1 LINK.
```

1. After the transaction processes, you can see the details for the complete the job run the **Runs** page in the Node Operators UI.

1. In Remix, click the `currentPrice` variable to see the current price updated on your consumer contract.

1. Input your operator contract address and the job ID for the `Get > String` job into the `requestFirstId` request method **without dashes**. The job ID is the `externalJobID` parameter, which you can find on your job's definition page in the Node Operators UI.

1. Click the **transact** button for the `requestFirstId` function and approve the transaction in Metamask. The `requestFirstId` function asks the node to retrieve the first `id` from [https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&per_page=10](https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&per_page=10).

1. After the transaction processes, you can see the details for the complete the job run the **Runs** page in the Node Operators UI.

1. In Remix, click the `id` variable to see the current price updated on your consumer contract.

> ### Check Forwarder

Reviewer: This section still TBD?

Author: @dwightjl yes , waiting for the next docker release

- Your node has two externally owned accounts (EOA).
- Your node has two [direct request](/chainlink-nodes/oracle-jobs/job-types/direct_request) jobs. One job returns _uint256_, and the other returns _string_.
- Each job uses a different EOA.
- You use a [forwarder](/chainlink-nodes/contracts/forwarder) contract to fulfill requests with two EOAs that look like a single address.

:::note
Here you are using a forwarder contract and two EOAs for two [direct request](/chainlink-nodes/oracle-jobs/job-types/direct_request) jobs. You can use the same strategy for different job types, VRF and OCR. Supporting different job types on the same Chainlink node reduces your infrastructure and maintenance costs.
:::

## Check your Chainlink node is running

On your terminal, check that your Chainlink node is running:

```shell
 docker ps -a -f name=chainlink

 CONTAINER ID   IMAGE                            COMMAND               CREATED         STATUS                   PORTS                    NAMES
 feff39f340d6   smartcontract/chainlink:1.12.0   "chainlink local n"   4 minutes ago   Up 4 minutes (healthy)   0.0.0.0:6688->6688/tcp   chainlink
```

> Your Chainlink Operator Interface is accessible on [http://localhost:6688](http://localhost:6688).

Reviewer: Should we include a reminder for how to create the SSH tunnel? I copy and paste the command from the docs frequently, so I find it very useful.

Author: good idea. check last commit

Metamask prompts you to confirm the transaction.

<ClickToZoom src='/images/chainlink-nodes/node-operators/forwarder/acceptAuthorizedReceivers.jpg' />

## Activate the forwarder

1. In the shell of your Chainlink node, enable the forwarder using the Chainlink CLI. Replace `forwarderAddress` with the forwarder address that was created by the factory. Replace `chainId` with the EVM chain ID. (`11155111` for Sepolia):

   ```shell
   chainlink forwarders track --address forwarderAddress --evmChainID chainId
   ```

   In this example, the command is:

   ```shell
   chainlink forwarders track --address 0xA3f07D6773514480b918C2742b027b3acD9E44fA --evmChainID 11155111
   ```

   ```shell
   Forwarder created
   ------------------------------------------------------
   ID:         1
   Address:    0xA3f07D6773514480b918C2742b027b3acD9E44fA
   Chain ID:   11155111
   Created At: 2023-03-02T11:41:43Zchainlink@22480bec8986
   ```

1. Exit the shell of your Chainlink node:

   ```shell
   exit
   ```

1. Stop your Chainlink node:

   ```shell
   docker stop chainlink && docker rm chainlink
   ```

> 1. Update your environment file. If you followed [Running a Chainlink Node locally](/chainlink-nodes/v1/running-a-chainlink-node) guide then add `ETH_USE_FORWARDERS=true` and `FEATURE_LOG_POLLER=true` to your environment file:

Reviewer: ```suggestion
1. Update your environment file. If you followed the [Running a Chainlink Node locally](/chainlink-nodes/v1/running-a-chainlink-node) guide, add `ETH_USE_FORWARDERS=true` and `FEATURE_LOG_POLLER=true` to your environment file:
```

> 1. Update your environment file. If you followed [Running a Chainlink Node locally](/chainlink-nodes/v1/running-a-chainlink-node) guide then add `ETH_USE_FORWARDERS=true` and `FEATURE_LOG_POLLER=true` to your environment file:

Reviewer: We'll need to update this for TOML format later, but for now .env is fine. Just a reminder.

Author: yes indeed.

   - `YOUR_OPERATOR_CONTRACT_ADDRESS` with the address of your deployed operator contract address.
   - `EOA_ADDRESS` with the **first** Chainlink node EOA.

     <CodeSample src="samples/ChainlinkNodes/forwarder/get-uint256.toml"/>

1. Create the `string` job. Replace:

   - `YOUR_OPERATOR_CONTRACT_ADDRESS` with the address of your deployed operator contract address.
   - `EOA_ADDRESS` with the **second** Chainlink node EOA.

     <CodeSample src="samples/ChainlinkNodes/forwarder/get-string.toml"/>

1. After you create the jobs, record the two job IDs.

:::note
Note that both jobs have the attribute `forwardingAllowed = true`. This attribute will make the jobs attempt to run through the forwarder before falling back to default mode. For example, if the Chainlink node tracks no forwarder, the job falls back to the default mode.
:::

## Test the transaction-sending strategy

### Create API requests

1. Open [APIConsumerForwarder.sol](https://remix.ethereum.org/#url=https://docs.chain.link/samples/APIRequests/APIConsumerForwarder.sol) in the Remix IDE.

1. Note that `setChainlinkToken(0x779877A7B0D9E8603169DdbD7836e478b4624789)` is configured for _Sepolia_.

> 1. On the **Compiler** tab, click the **Compile** button for `APIConsumerForwarder.sol`.

Reviewer: ```suggestion
1. On the **Compiler** tab, click the **Compile** button for `APIConsumerForwarder.sol`.
```

Previous sections use _italics_ when describing UI elements. Either bold or italics is fine, but we should be consistent. I recommend using bold ** for UI objects.

   <ClickToZoom src='/images/chainlink-nodes/node-operators/forwarder/factorycreatetransactions.jpg' />

1. Right-click on each contract address and open it in a new tab.
1. At this point, you should have one tab displaying the Operator contract and one tab displaying the Forwarder contract.

   <ClickToZoom src='/images/chainlink-nodes/node-operators/forwarder/operator.jpg' />
      <ClickToZoom src='/images/chainlink-nodes/node-operators/forwarder/forwarder.jpg' />

1. Record the operator and forwarder addresses. You will need them later.

## Access control setup

As explained in the [Forwarder](/chainlink-nodes/contracts/forwarder) page:

- The owner of a forwarder contract is an [operator](/chainlink-nodes/contracts/operator) contract. The owner of the operator contract is a more secure address, such as a hardware wallet or a multisig wallet. Therefore, node operators can manage a set of forwarder contracts through an operator contract using a secure account such as hardware or a multisig wallet.
- Forwarder contracts distinguish between owners and authorized senders. Authorized senders are hot wallets (Chainlink nodes' EOAs).

For this example to run, you will have to:

> - Allow the forwarder contract to call the operator's [fulfillOracleRequest2](/chainlink-nodes/contracts/operator#fulfilloraclerequest2) function by calling the [setauthorizedsenders](/chainlink-nodes/contracts/receiver#setauthorizedsenders) function on the operator contract. Specify the forwarder address as a parameter.

Reviewer 2: nit: capitalization not consistent

> - Allow the two Chainlink node EOAs to call the forwarder's [forward](/chainlink-nodes/contracts/forwarder#forward) function. Because the operator contract owns the forwarder contract, call [acceptAuthorizedReceivers](/chainlink-nodes/contracts/operator#acceptauthorizedreceivers) on the operator contract. Specify the forwarder contract address and the two Chainlink node EOAs as parameters. This call makes the operator contract accept ownership of the forwarder contract and authorizes the Chainlink node EOAs to call the forwarder contract by calling [setauthorizedsenders](/chainlink-nodes/contracts/receiver#setauthorizedsenders).

Reviewer 2: nit: capitalization not consistent
